sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfigTranslator.java

```diff
@@ -0,0 +1,345 @@
+      BigtableConfig config, PipelineOptions pipelineOptions) {
+    BigtableDataSettings.Builder dataBuilder;
+    if (!Strings.isNullOrEmpty(config.getEmulatorHost())) {
+      String hostAndPort = config.getEmulatorHost();
+      try {
+        int lastIndexOfCol = hostAndPort.lastIndexOf(":");
+        int port = Integer.parseInt(hostAndPort.substring(lastIndexOfCol + 1));
+        dataBuilder =
+            BigtableDataSettings.newBuilderForEmulator(
+                hostAndPort.substring(0, lastIndexOfCol), port);
+      } catch (NumberFormatException | IndexOutOfBoundsException ex) {
+        throw new RuntimeException("Invalid host/port in BigtableConfig " + hostAndPort);
+      }
+    } else {
+      dataBuilder = BigtableDataSettings.newBuilder();
+    }
+
+    // Configure target
+    dataBuilder
+        .setProjectId(Objects.requireNonNull(config.getProjectId().get()))
+        .setInstanceId(Objects.requireNonNull(config.getInstanceId().get()));
+    if (config.getAppProfileId() != null
+        && !Strings.isNullOrEmpty(config.getAppProfileId().get())) {
+      dataBuilder.setAppProfileId(Objects.requireNonNull(config.getAppProfileId().get()));
+    }
+
+    if (config.getCredentialFactory() != null) {
+      dataBuilder
+          .stubSettings()
+          .setCredentialsProvider(
+              FixedCredentialsProvider.create(
+                  ((GcpOptions) config.getCredentialFactory()).getGcpCredential()));
+    }
+
+    configureHeaderProvider(dataBuilder.stubSettings(), pipelineOptions);
+
+    return dataBuilder;
+  }
+
+  private static void configureHeaderProvider(
+      StubSettings.Builder<?, ?> stubSettings, PipelineOptions pipelineOptions) {
+
+    ImmutableMap.Builder<String, String> headersBuilder = ImmutableMap.<String, String>builder();
+    headersBuilder.putAll(stubSettings.getHeaderProvider().getHeaders());
+    headersBuilder.put(
+        GrpcUtil.USER_AGENT_KEY.name(), Objects.requireNonNull(pipelineOptions.getUserAgent()));
+
+    stubSettings.setHeaderProvider(FixedHeaderProvider.create(headersBuilder.build()));
+  }
+
+  private static BigtableDataSettings configureWriteSettings(
+      BigtableDataSettings.Builder settings, BigtableWriteOptions writeOptions) {
+    BigtableBatchingCallSettings.Builder callSettings =
+        settings.stubSettings().bulkMutateRowsSettings();
+    RetrySettings.Builder retrySettings = callSettings.getRetrySettings().toBuilder();
+    BatchingSettings.Builder batchingSettings = callSettings.getBatchingSettings().toBuilder();
+    if (writeOptions.getAttemptTimeout() != null) {
+      retrySettings.setInitialRpcTimeout(Duration.ofMillis(writeOptions.getAttemptTimeout()));
+
+      if (writeOptions.getOperationTimeout() == null) {
+        retrySettings.setTotalTimeout(
+            Duration.ofMillis(
+                Math.max(
+                    retrySettings.getTotalTimeout().toMillis(), writeOptions.getAttemptTimeout())));
```

Comment:
Do you need the if statement here? The next statement will clobber it, so I think you can skip the if here